Annotate the reference implementation for Audit

DanGould opened the pull request:
A high level, annotated "TODO" for integrating payjoin in your app. This info should make its way into the docs.rs documentation and payjoin.org. To start I've drafted plain ol' markdown so you can review and suggest edits to the doc directly.

It's a mental model of what the payjoin-client logic does and where/how the coin selection happens and the interplay with a wallet to understand how it could work with others like bdk.

This is a tool to try and figure out the bare-minimum shape of what you'd need as an API to use it on a mobile phone which has already access to e.g. bdk bindings

Thanks to @thunderbiscuit for making this request.

cc: @fiatjaf @kaloudis who also requested more detail on how to integrate the protocol

One of Kixunil's criteria for taking his original work into widespread production is getting an outside audit. This is a good start.

DanGould commented on MENTAL-MODEL.md:
````diff
@@ -0,0 +1,376 @@
+let link = link
+    .check_pj_supported()
+    .map_err(|e| anyhow!("The provided URI doesn't support payjoin (BIP78): {}", e))?;
+```
````
Refactoring ([see issue #19](https://github.com/payjoin/rust-payjoin/issues/19)) to use a builder pattern may eliminate the need to call `check_pj_supported` or expose bip21 at all.

DanGould commented on MENTAL-MODEL.md:
```diff
@@ -0,0 +1,376 @@
+    .wallet_process_psbt(&psbt, None, None, None)
+    .with_context(|| "Failed to process PSBT")?
+    .psbt;
+let psbt = load_psbt_from_base64(psbt.as_bytes()) // SHOULD BE PROVIDED BY CRATE AS HELPER USING rust-bitcoin base64 feature
```
The `load_psbt_from_base64` helper method could be refactored to use the `bitcoin::base64` feature. I imagine the rust-bitcoin crate would be open to `Psbt::to_base64` and `Psbt::from_base64` feature-gated helpers being added as well. see https://github.com/payjoin/rust-payjoin/issues/30

DanGould commented on MENTAL-MODEL.md:
````diff
@@ -0,0 +1,376 @@
+    payjoin::bitcoin::Amount::from_sat(10000),
+    None,
+);
+```
````
I think the current flow works and **it does not need to change to prove the concept**, but it may also be improved with the builder pattern (#19). The builder might take the bip21, an optional amount that differs from that in the bip21, fee rate, a closure to fund and sign an Original PSBT. The result should be a request containing `(original_psbt, optional_parameters)` and state in context.

DanGould commented on MENTAL-MODEL.md:
```diff
@@ -0,0 +1,376 @@
+
+Senders request a payjoin from the receiver with a payload containing the Original PSBT  and optional parameters. They require a secure endpoint for authentication and message secrecy to prevent that transaction from being modified by a malicious third party during transit or being snooped on. Only https and .onion endpoints are spec-compatible payjoin endpoints.
+
+Avoiding the secure endpoint requirement is convenient for testing.
```
(but again, Signal guidelines remind me of the folly of options). I wonder if there's a better way to set up a testing environment with a valid https cert with something like Nix.

DanGould commented on MENTAL-MODEL.md:
````diff
@@ -0,0 +1,376 @@
+log::debug!("Proposed psbt: {:#?}", psbt);
+```
+
+Payjoin response errors (called [receiver's errors in spec](https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki#receivers-well-known-errors)) come from a remote server and can be used to "maliciously to phish a non technical user." Only those listed as "well known" in the spec should be displayed with preset messages to prevent phishing.
````
Errors from `process_response` may be improved by being separated into safe `ReceiverError::WellKnown` standard error types that can be displayed in the UI and `ReceiverError::DebugOnly` which "can only appear in debug logs." Separation would simplify an integration's error handling. (https://github.com/payjoin/rust-payjoin/issues/53)

DanGould commented on MENTAL-MODEL.md:
````diff
@@ -0,0 +1,376 @@
+    .with_context(|| "Failed to finalize PSBT")?
+    .hex
+    .ok_or_else(|| anyhow!("Incomplete PSBT"))?;
+```
````
The signature step could be combined with the process step if a signing closure were passed to process_response. That may error if the closure is provided from an async runtime, however

DanGould commented on MENTAL-MODEL.md:
```diff
@@ -0,0 +1,376 @@
+fn receive_payjoin(
+    bitcoind: bitcoincore_rpc::Client,
+    amount_arg: &str,
+    endpoint_arg: &str,
```
endpoint displayed in the BIP 21 URI could be configured rather than passed as an argument, and the amount should be optional https://github.com/payjoin/rust-payjoin/issues/54

DanGould commented on MENTAL-MODEL.md:
````diff
@@ -0,0 +1,376 @@
+let _pj_uri = pj_uri
+    .check_pj_supported()
+    .map_err(|e| anyhow!("Constructed URI does not support payjoin: {}", e))?;
+```
````
A URI builder may be a more ergonomic interface. The `bip21` crate might be a good candidate to start to explore bindings with. It doesn't make sense to construct a string and then parse it rather than provide a concrete `PjUri` constructor taking `address: Address`, `amount: Option<amount>`, `pj: Endpoint`

DanGould commented on MENTAL-MODEL.md:
````diff
@@ -0,0 +1,376 @@
+Start a server to respond to payjoin protocol POST messages.
+
+```rs
+rouille::start_server("0.0.0.0:3000", move |req| handle_web_request(&req, &bitcoind));
````
The listening port should be configurable in the reference implementation. This is distinct from the `pj` endpoint which is where this port is exposed to the outside internet.

DanGould commented on MENTAL-MODEL.md:
```diff
@@ -0,0 +1,376 @@
+rouille::start_server("0.0.0.0:3000", move |req| handle_web_request(&req, &bitcoind));
+// ...
+fn handle_web_request(req: &Request, bitcoind: &bitcoincore_rpc::Client) -> Response {
+    handle_payjoin_request(req, bitcoind)
+        .map_err(|e| match e {
+            ReceiveError::RequestError(e) => {
+                log::error!("Error handling request: {}", e);
+                Response::text(e.to_string()).with_status_code(400)
+            }
+            e => {
+                log::error!("Error handling request: {}", e);
+                Response::text(e.to_string()).with_status_code(500)
+            }
+        })
```
It would be a better experience to respond to HTTP GET requests to this endpoint from browsers with a payjoin QuickStart HTML page instead of 404 or 500 errors.

DanGould commented on MENTAL-MODEL.md:
````diff
@@ -0,0 +1,376 @@
+        copy.find(|(k, _)| k.eq_ignore_ascii_case(key)).map(|(_, v)| v)
+    }
+}
+```
````
I reckon bindings will run into a problem with this Trait-based design. @thunderbiscuit, what do you think?

DanGould commented on MENTAL-MODEL.md:
````diff
@@ -0,0 +1,376 @@
+Check the sender's Original PSBT to prevent it from stealing, probing to fingerprint its wallet, and to avoid privacy gotchas.
+
+```rs
+// in a payment processor where the sender could go offline, this is where you schedule to broadcast the original_tx
+let _to_broadcast_in_failure_case = proposal.get_transaction_to_schedule_broadcast();
````
The broadcast out of order! This should be after check 1 making sure it's broadcastable.

DanGould commented on MENTAL-MODEL.md:
````diff
@@ -0,0 +1,376 @@
+}
+```
+
+Serious, in-depth research has gone into proper transaction construction. [Here's a good starting point from the JoinMarket repo](https://gist.github.com/AdamISZ/4551b947789d3216bacfcb7af25e029e#gistcomment-2796539). Using methods for coin selection not provided by this library may have dire implications for privacy.
````
Downstream projects may benefit to consume expose the selection algorithm inside `try_preserving_privacy` as a static function. It may be easier easier and more lightweight to bind to and avoid dangerous instances of UIH in addition to the typestate method. Look at BDK Coin Selection for interface inspiration.

DanGould replied:
https://github.com/payjoin/rust-payjoin/issues/70

DanGould commented on MENTAL-MODEL.md:
````diff
@@ -0,0 +1,376 @@
+Fees are applied to the augmented Payjoin Proposal PSBT using calculation factoring the receiver's own preferred feerate and the sender's fee-related [optional parameters](https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki#optional-parameters). The current `apply_fee` method is primitive, disregarding PSBT fee estimation and only adding fees coming from the sender's budget. When more  ...
+
+```rs
+let payjoin_proposal_psbt = payjoin.apply_fee(min_feerate_sat_per_vb: Some(1))?;
````
The current `apply_fee` method is primitive, disregarding fee estimation based on PSBT contents and only adding fees coming from the sender's budget. When more accurate tools are available to calculate a PSBT's fee-dependent weight (slightly more complicated than it sounds, but solved, just unimplemented in rust-bitcoin), this `apply_fee` should be improved.

DanGould commented on MENTAL-MODEL.md:
````diff
@@ -0,0 +1,376 @@
+```rs
+let payjoin_proposal_psbt = payjoin.apply_fee(min_feerate_sat_per_vb: Some(1))?;
+
+log::debug!("Extracted PSBT: {:#?}", payjoin_proposal_psbt);
+// Sign payjoin psbt
+let payjoin_base64_string =
+    base64::encode(bitcoin::consensus::serialize(&payjoin_proposal_psbt));
+// `wallet_process_psbt` adds available utxo data and finalizes
+let payjoin_proposal_psbt =
+    bitcoind.wallet_process_psbt(&payjoin_base64_string, sign: None, sighash_type: None, bip32derivs: Some(false))?.psbt;
+let payjoin_proposal_psbt =
+    load_psbt_from_base64(payjoin_proposal_psbt.as_bytes()).context("Failed to parse PSBT")?;
+```
````
The signing algorithm could be passed as a closure, but may encounter runtime issues as mentioned previously. I.e., rust does not yet support async closures.

DanGould replied:
While bitcoind's `wallet_process_psbt` handles most everything you throw at it with grace, LND's PSBT signing [gets very cranky](https://github.com/lightningnetwork/lnd/issues/7266) when fed inputs belonging to other wallets and requires [a complicated workaround of multiple gRPC calls](https://github.com/chaincase-app/nolooking/blob/a2be2adbf2f75074e9e0d913f761e834b8ce70d1/src/scheduler.rs#L367-L410) to function properly. It might be good to note this in the documentation since it's such a pain in the kiester.

cryptoquick commented on MENTAL-MODEL.md:
````diff
@@ -0,0 +1,376 @@
+// TODO display well-known errors and log::debug the rest
+let psbt = ctx.process_response(response).with_context(|| "Failed to process response")?;
+log::debug!("Proposed psbt: {:#?}", psbt);
+```
````
Where is ctx from? Please try to make examples more complete.

DanGould replied:
ctx is the context returned from create_pj_request in step 4, so I've added a comment. Would it be better to have a complete send_payjoin method based on a file in examples rather than this breakdown, or something else?

cryptoquick reviewed (changes requested):
Just some suggestions. I kind of ran out of steam near the end there, it's a pretty lengthy document. Hopefully what I was able to provide was helpful.

DanGould commented:
I've moved this into rustdoc for `send` and `receive` features. I've adjusted the example to take into consideration feedback and changes for 0.8.0. I plan to send it with that update tomorrow unless some major error is found in review.

The original audit doc can be found here: https://gist.github.com/DanGould/595865688cabcfec42eb0e2a36f65490
